### PR TITLE
Support booting VMs from a custom kernel/initrd images with custom kernel arguments

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
This feature allows to define a custom kernel image / initrd image for a VMI to boot from. It also allows to define custom kernel command line arguments to be passed directly to the installer. More info can be found [here](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/virtualization_administration_guide/sub-sect-op-sys-dir-kern-boot).

This can be define as following:
```
spec:
    firmware:
      kernelBoot:
        kernelArgs: console=ttyS0 <mandatory>
        container: <optional>
          image: registry:5000/kubevirt/alpine-ext-kernel-boot-demo:devel <obligatory if container defined>
          initrdPath: /boot/initramfs-virt <optional>
          kernelPath: /boot/vmlinuz-virt <optional>
          imagePullSecret: .. <optional>
          imagePullPolicy: .. <optional>
```

The reason for `container` to be optional is that we might allow ways for providing the images in the future.

In short, the implementation flow is:
1. virt-controller, in `template.go`, checks if a container is specified there and adds it to the pod template.
2. virt-handler has in `pkg/virt-handler/container-disk/mount.go` more mount/unmount function for kernel artifacts (which are used by `pkg/virt-handler/vm.go` before or after Mount for the containerdisks is called, if a container is specified).
3. virt-launcher converts the specified kernel boot definition into libvirt's domain definitions.

**Special notes for your reviewer**:
This PR is the continuation of [this work](https://github.com/kubevirt/kubevirt/pull/3613) which is discontinued. The implementation was at a very early phase but the comments and discussions there were very helpful and important.

**Release note**:

```release-note
[feature] support booting VMs from a custom kernel/initrd images with custom kernel arguments
```

### CODE REVIEW

File: pkg/virt-handler/container-disk/mount.go [hunk 1]
```diff
@@ -367,3 +363,199 @@ func (m *mounter) ContainerDisksReady(vmi *v1.VirtualMachineInstance, notInitial
...
+			// cleaning the mounted files.
+			log.Log.Object(vmi).Reason(err).Error("unable to unmount kernel artifacts")
+		}
+
+		err = os.Remove(targetDir)
+		if err != nil {
+			log.DefaultLogger().Object(vmi).Infof("cannot delete dir %s. err: %v", targetDir, err)
+		}
+
+		removeSliceElement := func(s []vmiMountTargetEntry, idxToRemove int) []vmiMountTargetEntry {
```
Comment (iholder-redhat) [inline pkg/virt-handler/container-disk/mount.go:533]: I have read that this is the fastest "go way" to remove an element from a slice.. is there a better way?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-update-admitter.go [hunk 2]
```diff
@@ -82,7 +82,7 @@ func admitHotplug(newVolumes, oldVolumes []v1.Volume, newDisks, oldDisks []v1.Di
 		return webhookutils.ToAdmissionResponse([]metav1.StatusCause{
 			{
 				Type:    metav1.CauseTypeFieldValueInvalid,
-				Message: "number of disks does not equal the number of volumes",
+				Message: fmt.Sprintf("number of disks (%d) does not equal the number of volumes (%d)", len(newDisks), len(newVolumes)),
```
Comment (vladikr) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-update-admitter.go:85]: I'm just curious, is this change directly related to this PR?

File: pkg/container-disk/container-disk.go [hunk 3]
```diff
@@ -176,66 +213,113 @@ func GenerateContainers(vmi *v1.VirtualMachineInstance, podVolumeName string, bi
...
 }
 
+func GenerateKernelBootContainer(vmi *v1.VirtualMachineInstance, podVolumeName string, binVolumeName string) *kubev1.Container {
+	if !util.IsKernelBootDefinedProperly(vmi) {
+		if f := vmi.Spec.Domain.Firmware; f != nil && f.KernelBoot != nil && f.KernelBoot.Container != nil &&
+			f.KernelBoot.Container.InitrdPath == "" && f.KernelBoot.Container.KernelPath == "" {
+			log.Log.Object(vmi).Warning("Kernel boot container does not contain kernel path nor initrd path - ignoring kernel-boot configuration")
+		}
+		return nil
+	}
```
Comment (vladikr) [inline pkg/container-disk/container-disk.go:223]: From my point of view, this validation should belong in the validation webhook[1] and we should be rejecting incorrect specs, instead of ignoring them. 
But I wonder what is your take on this?

[1] pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

File: pkg/container-disk/container-disk.go [hunk 4]
```diff
@@ -176,66 +213,113 @@ func GenerateContainers(vmi *v1.VirtualMachineInstance, podVolumeName string, bi
...
+
+	kernelBootContainer := vmi.Spec.Domain.Firmware.KernelBoot.Container
+
+	kernelBootVolume := v1.Volume{
+		Name: KernelBootVolumeName,
+		VolumeSource: v1.VolumeSource{
+			ContainerDisk: &v1.ContainerDiskSource{
+				Image:           kernelBootContainer.Image,
+				ImagePullSecret: kernelBootContainer.ImagePullSecret,
+				Path:            "/",
```
[the remark this answers is not in the document]
Comment (vladikr) [reply pkg/container-disk/container-disk.go:233]: It's an interesting suggestion. I wonder though if we shouldn't just go with the simple solution in the first iteration...

What do you think would be a use case for having such a nested dir structure for a container kernel boot image?
I would assume that the vast majority of users will just have these files in something like /boot.

File: pkg/virt-handler/container-disk/mount.go [hunk 5]
```diff
@@ -26,18 +27,21 @@ import (
...
-	mountRecordsLock       sync.Mutex
-	suppressWarningTimeout time.Duration
-	pathGetter             containerdisk.SocketPathGetter
+	podIsolationDetector       isolation.PodIsolationDetector
+	mountStateDir              string
+	mountRecords               map[types.UID]*vmiMountTargetRecord
+	mountRecordsLock           sync.Mutex
+	suppressWarningTimeout     time.Duration
+	SocketPathGetter           containerdisk.SocketPathGetter
+	KernelBootSocketPathGetter containerdisk.KernelBootSocketPathGetter
```
Comment (vladikr) [inline pkg/virt-handler/container-disk/mount.go:36]: Do you need these attributes to be public? or can it be kernelBootSocketPathGetter and socketPathGetter?

File: pkg/virt-handler/container-disk/mount.go [hunk 6]
```diff
@@ -367,3 +363,199 @@ func (m *mounter) ContainerDisksReady(vmi *v1.VirtualMachineInstance, notInitial
...
+	if isMounted, err := areKernelArtifactsMounted(targetDir, targetInitrdPath, targetKernelPath); err != nil {
+		return fmt.Errorf("failed to determine if %s is already mounted: %v", targetDir, err)
+	} else if !isMounted {
+		log.Log.Object(vmi).Infof("mounting kernel artifacts are not mounted - mounting...")
+		rootPath, err := m.getRootPathFromSocket(vmi, socketFilePath, kernelBootName, nodeRes)
+		if err != nil {
+			return fmt.Errorf("failed to detect root mount point of %v on the node: %v", kernelBootName, err)
+		}
+
+		err = os.Mkdir(targetDir, 777)
```
[the remark this answers is not in the document]
Comment (vladikr) [reply pkg/virt-handler/container-disk/mount.go:428]: I see that 0755 is used everywhere else in virt-handler's mount, we can be consistent with it.

File: pkg/virt-handler/container-disk/mount.go [hunk 7]
```diff
@@ -367,3 +363,199 @@ func (m *mounter) ContainerDisksReady(vmi *v1.VirtualMachineInstance, notInitial
...
+		} else if !mounted {
+			return fmt.Errorf("kernel artifacts verification failed")
+		}
+	}
+
+	return nil
+}
+
+func (m *mounter) UnmountKernelArtifacts(vmi *v1.VirtualMachineInstance) error {
+	if !util.IsKernelBootDefinedProperly(vmi) || vmi.UID == "" {
```
[the remark this answers is not in the document]
Comment (vladikr) [reply pkg/virt-handler/container-disk/mount.go:478]: I wouldn't worry about this. I see that this is being called from `processVmCleanup` and I don't see how vmi wouldn't have a UID at this stage.

**Review by vladikr — COMMENTED** [review]
Thanks for putting this PR together!

Overall it looks good. I've made some comments.
We should also check that the VMI is live migratable.
We already have this test[1] with a bunch of disks that we are testing the migration with. Perhaps you could add addKernelBootContainer there as well and make sure it works.

[1] https://github.com/kubevirt/kubevirt/blob/master/tests/migration_test.go#L1795

File: pkg/container-disk/container-disk.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply pkg/container-disk/container-disk.go:233]: Well, theoretically the limitation for container-disks (have to point to a directory with exactly one file) does not exist and it's possible, but you're probably right and I'm over-thinking it.
I agree, let's go with the simple solution for now.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-update-admitter.go (hunk 2 above)
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-update-admitter.go:85]: Actually it isn't. Sorry, next time I'll point it out in a comment here.
It's just that I changed this for debugging then thought it's better to stay that way. Opening a PR especially for that seemed a bit extreme to me.

File: pkg/container-disk/container-disk.go (hunk 3 above)
Comment (iholder-redhat) [reply pkg/container-disk/container-disk.go:223]: I need to fully understand something before answering that. Take a look on [this line of code](https://github.com/kubevirt/kubevirt/blob/1c9285807dac4684a3d3b0d77c0b94bc04a30609/pkg/virt-launcher/virtwrap/converter/converter.go#L1163). I'm converting the command line arguments to libvirt domain even if kernel/initrd images are not provided. From Roman's discussion on the previous PR I had the impression that this is the right way to go but now I'm re-thinking it.

Do command line arguments have effect even if no kernel/initrd images are provided? If there is an effect than a spec with only command line should be valid and command line args will be converted into the domain (sine `util.IsKernelBootDefinedProperly()` will return true). The validation you mentioned happens only in a single place before generating the container-disk (so that if kernel/initrd images aren't defined no containers will be generated).

If this has no effect though then I agree - the spec should be considered invalid and we can enforce that with validation webhook.

P.S. is it valid to provide initrd image only (without a kernel image)?

File: pkg/virt-handler/container-disk/mount.go (hunk 5 above)
Comment (iholder-redhat) [reply pkg/virt-handler/container-disk/mount.go:36]: good catch, thanks!

File: pkg/util/util.go [hunk 8]
```diff
@@ -69,3 +69,21 @@ func ResourceNameToEnvVar(prefix string, resourceName string) string {
...
+	if vmi == nil {
+		return false
+	}
+
+	vmiFirmware := vmi.Spec.Domain.Firmware
+	if (vmiFirmware == nil) || (vmiFirmware.KernelBoot == nil) || (vmiFirmware.KernelBoot.Container == nil) {
+		return false
+	}
+
+	if c := vmiFirmware.KernelBoot.Container; (c.KernelPath == "") && (c.InitrdPath == "") {
```
Comment (vladikr) [inline pkg/util/util.go:84]: I think it should be fine to specify initrd without the kernel image - I don't see that libvirt nor qemu prevents this.

File: pkg/container-disk/container-disk.go (hunk 3 above)
Comment (vladikr) [reply pkg/container-disk/container-disk.go:223]: Sorry, I shouldn't do reviews so late ;)


> I need to fully understand something before answering that. Take a look on [this line of code](https://github.com/kubevirt/kubevirt/blob/1c9285807dac4684a3d3b0d77c0b94bc04a30609/pkg/virt-launcher/virtwrap/converter/converter.go#L1163). I'm converting the command line arguments to libvirt domain even if kernel/initrd images are not provided. From Roman's discussion on the previous PR I had the impression that this is the right way to go but now I'm re-thinking it.

It is, afaik, I don't see any limitation here. 

> Do command line arguments have effect even if no kernel/initrd images are provided? If there is an effect than a spec with only command line should be valid and command line args will be converted into the domain (sine `util.IsKernelBootDefinedProperly()` will return true). The validation you mentioned happens only in a single place before generating the container-disk (so that if kernel/initrd images aren't defined no containers will be generated).

What I meant was that we shouldn't accept specs that are malformed...
For example
```
spec:
    firmware:
      kernelBoot:
        container:
          image: myimage/ext-kernel-boot-demo
```
Would we know where to take the kernel and initrd from this image?

If I read the code correctly, then you simply suggest to skip this container image, while I think we should explicitly reject this.


> P.S. is it valid to provide initrd image only (without a kernel image)?

afaik, yes.

Comment (vladikr) [reply pkg/container-disk/container-disk.go:223]: I would also rename `IsKernelBootDefinedProperly` to something like `HasKernelBootContainerImage` or `ShouldUseKernelBootContainerImage` or differently, just in a way that it will indicate whether we are going to use this container disk or not.

Comment (iholder-redhat) [reply pkg/container-disk/container-disk.go:223]: I see your point. What went through my head is that someone might use the following configuration:

```
spec:
    firmware:
      kernelBoot:
        commandLine: "something valid"   
        container:
          image: myimage/ext-kernel-boot-demo
```

This definition is, as you claimed, invalid as an image is defined without a path to kernel/initrd images. But this definition is also partially valid in the sense that command line arguments can still be used.

As I see it in this situation we have 2 options - either to fail with an error or use the partially valid configuration (command-line args) and trigger a warning about the problematic definition which will be ignored. I thought the the later option is better then the first.

What do you think?

Comment (iholder-redhat) [reply pkg/container-disk/container-disk.go:223]: And I agree, `IsKernelBootDefinedProperly` bugs me too. After we finish this discussion I will choose a better name :)

File: pkg/util/util.go (hunk 8 above)
Comment (iholder-redhat) [reply pkg/util/util.go:84]: You're absolutely right.
This condition is true (which means that `false` will be returned == kernel boot is not defined properly) only if both kernel and initid are not defined.

File: pkg/container-disk/container-disk.go (hunk 3 above)
Comment (vladikr) [reply pkg/container-disk/container-disk.go:223]: > I see your point. What went through my head is that someone might use the following configuration:
> 
> ```
> spec:
>     firmware:
>       kernelBoot:
>         commandLine: "something valid"   
>         container:
>           image: myimage/ext-kernel-boot-demo
> ```
> 
> This definition is, as you claimed, invalid as an image is defined without a path to kernel/initrd images. But this definition is also partially valid in the sense that command line arguments can still be used.
> 
> As I see it in this situation we have 2 options - either to fail with an error or use the partially valid configuration (command-line args) and trigger a warning about the problematic definition which will be ignored. I thought the the later option is better then the first.
> 
> What do you think?

My take on this is that the validating webhook allows us to be explicit.
Rejecting such specs early allows us to reduce the amount of "guessing" we have to do to understand the original intent of the user.
The user then can quickly fix the issue and post an updated spec.

Alternatively, If we will just ignore a malformed container section, while the user expects to boot from this container image - he will spend a lot of time going through all the logs looking for what went wrong...
I guess I would prefer to avoid this and notify the user that something is wrong ahead of time.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go [hunk 9]
```diff
@@ -2225,3 +2226,31 @@ func validateDisks(field *k8sfield.Path, disks []v1.Disk) []metav1.StatusCause {
 
 	return causes
 }
+
+// Rejects kernel boot defined with initrd/kernel path but without an image
+func validateKernelBoot(field *k8sfield.Path, kernelBoot *v1.KernelBoot) (causes []metav1.StatusCause) {
```
Comment (vladikr) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:2231]: Let's add a unit test for this. I know you have a functional test, but it doesn't contradict.

**Review by vladikr — COMMENTED** [review]
@iholder-redhat great work!
It looks great right now. Thanks for making the changes.
Let's just add the below unit test and it will be good to go from my point of view.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go (hunk 9 above)
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:2231]: Done, thanks!

**Comment by vladikr** [comment]
Looks good to me now!
Thanks!
Let's see that these failed tests are not related... 
/approve

**Comment by vladikr** [comment]
/hold
@iholder-redhat  can you please look at the generate failures?

**Comment by iholder-redhat** [comment]
> /hold
> @iholder-redhat can you please look at the generate failures?

@vladikr It was due to the rebase with [this PR](https://github.com/kubevirt/kubevirt/pull/5495). I ran `make generate` (the new `examples/vmi-kernel-boot.yaml` file is slightly changed), should be fine now.
